File: tests/visitlog.py
# pub/visitlog.js, the browser half of counting the first visit (R-023; backend migrations 067-068), in a real browser
# with every Supabase request intercepted: nothing reaches production, and a request to anything but log_first_visit or
# public_partners fails the run. Checks what leaves the browser (and what never does), the visit id, where the person came
# from, the 60-a-visit cap, the keepalive request for leaving, Global Privacy Control, the sandbox, and no network.
#   python3 -m http.server 8832   (in this folder, once)
#   python3 tests/visitlog.py
import json, re, sys
from playwright.sync_api import sync_playwright

BASE = 'http://localhost:8832'
passed = failed = 0
def ok(cond, msg):
    global passed, failed
    print(('PASS ' if cond else 'FAIL ') + msg)
    if cond: passed += 1
    else: failed += 1

def rig(ctx, sent, partners=None, fail_partners=False):
    def rpc(route):
        req = route.request
        body = json.loads(req.post_data or '{}')
        sent.append({'url': req.url, 'headers': req.headers, 'p': body.get('p'), 'method': req.method})
        route.fulfill(status=204, body='', headers={'access-control-allow-origin': '*'})
    ctx.route(re.compile(r'.*supabase\.co/rest/v1/rpc/log_first_visit.*'), rpc)
    def pp(route):
        sent.append({'url': route.request.url, 'partners': True})
        if fail_partners: route.abort(); return
        m = re.search(r'slug=eq\.([a-z0-9-]+)', route.request.url)
        rows = [r for r in (partners or []) if m and r['slug'] == m.group(1)]
        route.fulfill(status=200, content_type='application/json', body=json.dumps(rows[0] if rows and 'vnd.pgrst.object' in (route.request.headers.get('accept') or '') else rows),
                      headers={'access-control-allow-origin': '*'})
    ctx.route(re.compile(r'.*supabase\.co/rest/v1/public_partners.*'), pp)
    # anything else on supabase.co would be a leak: refuse it and note it
    def other(route):
        sent.append({'url': route.request.url, 'leak': True}); route.abort()
    ctx.route(re.compile(r'.*supabase\.co/(?!rest/v1/(rpc/log_first_visit|public_partners)).*'), other)

IMPORT = "async () => { window.VL = await import('/pub/visitlog.js'); return true; }"
with sync_playwright() as pw:
    br = pw.chromium.launch()
    # ---- 1. a partner link from Instagram, on a phone ----
    ctx = br.new_context(viewport={'width': 390, 'height': 844})
    sent = []; logs = []
    rig(ctx, sent, partners=[{'slug': 'kokua-kalihi', 'welcome': 'Welcome, friends of Kokua Kalihi Valley!'}])
    page = ctx.new_page(); page.on('console', lambda m: logs.append((m.type, m.text)))
    page.goto(BASE + '/tests/?via=Kokua-Kalihi&utm_source=Instagram&utm_medium=social&utm_campaign=Fall%202026', referer='https://l.instagram.com/?u=https%3A%2F%2Fexample')
    page.evaluate(IMPORT)
    r1 = page.evaluate("() => VL.logVisit('topics', 'view', { path: 'in', seconds: 0.4 })")
    r2 = page.evaluate("() => VL.logVisit('issues', 'next', { path: 'in', seconds: 40.6, counts: { cats: 2, issues: 3, stances: 99, name: 'Leilani', email: 'yes' }, issue_ids: ['a1', 'b2', 'a1', 5], email: 'someone@example.com', name: 'Leilani' })")
    r3 = page.evaluate("() => VL.logVisit('topics', 'next', { issue_ids: ['a1'], quiz: 'maybe', lesson_step: 12.2, seconds: 99999 })")
    page.wait_for_timeout(800)
    rpc = [s for s in sent if s.get('p')]
    ok(r1 is True and r2 is True and r3 is True, f'logVisit resolves true when sent ({r1}, {r2}, {r3})')
    ok(len(rpc) == 3, f'three events, three requests ({len(rpc)})')
    p0 = rpc[0]['p'] if rpc else {}
    ok(re.match(r'^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$', p0.get('visit', '')) is not None, 'a random v4 visit id: ' + p0.get('visit', ''))
    ok(all(s['p']['visit'] == p0.get('visit') for s in rpc), 'the same id on every event of the visit')
    ok(p0.get('via') == 'kokua-kalihi' and p0.get('utm_source') == 'instagram' and p0.get('utm_medium') == 'social', 'via and utm words, lower-cased: ' + json.dumps({k: p0.get(k) for k in ('via', 'utm_source', 'utm_medium')}))
    ok('utm_campaign' not in p0, 'a utm word with a space is dropped, not sent')
    ok(p0.get('ref_domain') == 'instagram.com', 'the referring site by name only, l.instagram.com -> instagram.com: ' + str(p0.get('ref_domain')))
    ok(p0.get('device') == 'phone', 'a 390px window is a phone')
    ok(p0.get('seconds') == 0 and p0.get('path') == 'in', 'seconds rounded, path kept')
    p1 = rpc[1]['p'] if len(rpc) > 1 else {}
    ok(set(p1.keys()) <= {'visit', 'step', 'event', 'via', 'utm_source', 'utm_medium', 'ref_domain', 'device', 'path', 'seconds', 'counts', 'issue_ids'}, 'only the allowed keys leave the browser: ' + ','.join(sorted(p1.keys())))
    ok('email' not in p1 and 'name' not in p1 and 'Leilani' not in json.dumps(p1) and 'someone@' not in json.dumps(p1), 'never an email or a name, even when a caller passes one')
    ok(p1.get('counts') == {'cats': 2, 'issues': 3}, 'counts: known keys only, a number out of range dropped: ' + json.dumps(p1.get('counts')))
    ok(p1.get('issue_ids') == ['a1', 'b2'] and p1.get('seconds') == 41, 'issue ids once each, strings only; seconds rounded')
    p2 = rpc[2]['p'] if len(rpc) > 2 else {}
    ok('issue_ids' not in p2 and 'quiz' not in p2 and 'lesson_step' not in p2 and p2.get('seconds') == 3600, 'issue ids only with the issue screen\'s Next; a bad answer and a lesson step of 12 dropped; seconds held to an hour')
    ok(all('authorization' not in {k.lower() for k in s['headers']} or s['headers'].get('authorization', '').startswith('Bearer ') for s in rpc), 'no account is attached beyond what supabase-js sends')
    ok(page.evaluate("() => VL.visitVia()") == 'kokua-kalihi', 'visitVia() gives the slug')
    # ---- leaving: a keepalive request straight to the endpoint ----
    n0 = len(rpc)
    page.evaluate("() => VL.logVisit('stand', 'leave', { seconds: 5 })")
    page.wait_for_timeout(500)
    rpc = [s for s in sent if s.get('p')]
    lv = rpc[-1] if len(rpc) > n0 else {}
    ok(lv.get('p', {}).get('event') == 'leave' and lv.get('url', '').endswith('/rest/v1/rpc/log_first_visit') and lv.get('headers', {}).get('apikey', '').startswith('sb_publishable_'),
       'leave goes as its own request with the public key only')
    ok('authorization' not in {k.lower() for k in lv.get('headers', {})}, 'and no account')
    # ---- the same visit after a reload; a new tab is a new visit ----
    vid = p0.get('visit')
    page.reload(); page.evaluate(IMPORT)
    page.evaluate("() => VL.logVisit('stand', 'view')"); page.wait_for_timeout(400)
    rpc = [s for s in sent if s.get('p')]
    ok(rpc[-1]['p']['visit'] == vid and rpc[-1]['p'].get('via') == 'kokua-kalihi', 'a reload keeps the visit and where it came from')
    p2page = ctx.new_page(); p2page.goto(BASE + '/tests/'); p2page.evaluate(IMPORT)
    p2page.evaluate("() => VL.logVisit('topics', 'view')"); p2page.wait_for_timeout(400)
    rpc = [s for s in sent if s.get('p')]
    ok(rpc[-1]['p']['visit'] != vid and 'via' not in rpc[-1]['p'] and 'ref_domain' not in rpc[-1]['p'], 'a new tab is a new visit, with nothing carried over')
    # ---- the cap: 60 a visit ----
    before = len([s for s in rpc if s['p']['visit'] == vid])
    res = page.evaluate("async () => { const r = []; for (let k = 0; k < 70; k++) r.push(await VL.logVisit('session', 'next', { lesson_step: k % 10 })); return r.filter(Boolean).length; }")
    page.wait_for_timeout(1200)
    mine = len([s for s in sent if s.get('p') and s['p']['visit'] == vid])
    ok(mine == 60, f'a visit sends 60 events at most ({before} before, {res} of 70 more sent, {mine} in all)')
    # ---- a wrong step name is a programming error: warned, never sent ----
    k0 = len(sent)
    bad = p2page.evaluate("() => VL.logVisit('tour', 'view')"); p2page.wait_for_timeout(200)
    ok(bad is False and len(sent) == k0, 'an unknown step is not sent')
    # ---- partners ----
    w1 = page.evaluate("() => VL.partnerWelcome('kokua-kalihi')"); w2 = page.evaluate("() => VL.partnerWelcome('kokua-kalihi')")
    asked = len([s for s in sent if s.get('partners')])
    ok(w1 == 'Welcome, friends of Kokua Kalihi Valley!' and w2 == w1 and asked == 1, f'partnerWelcome reads public_partners once per slug ({asked} request): {w1}')
    ok(page.evaluate("() => VL.partnerWelcome('nobody-here')") is None, 'an unknown partner: null')
    ok(page.evaluate("() => VL.partnerWelcome('Bad Slug!')") is None and len([s for s in sent if s.get('partners')]) == 2, 'a bad slug: null, without asking')
    ok(not [s for s in sent if s.get('leak')], 'no other request reached Supabase: ' + json.dumps([s['url'] for s in sent if s.get('leak')]))
    ok(not [l for l in logs if l[0] == 'error'], 'no console errors: ' + json.dumps([l for l in logs if l[0] == 'error'])[:300])
    ctx.close()

    # ---- 2. Global Privacy Control: nothing at all ----
    ctx = br.new_context(viewport={'width': 1440, 'height': 900})
    ctx.add_init_script("Object.defineProperty(navigator, 'globalPrivacyControl', { get: () => true })")
    sent = []; rig(ctx, sent)
    page = ctx.new_page(); page.goto(BASE + '/tests/?via=kokua-kalihi'); page.evaluate(IMPORT)
    r = page.evaluate("() => VL.logVisit('topics', 'view', { path: 'in' })"); page.wait_for_timeout(400)
    stored = page.evaluate("() => sessionStorage.getItem('hiphi_fv')")
    ok(r is False and not [s for s in sent if s.get('p')] and stored is None, 'with Global Privacy Control nothing is sent and nothing is kept')
    ok(page.evaluate("() => VL.visitVia()") == 'kokua-kalihi', 'but the welcome line still knows the partner (visitVia reads the address)')
    ctx.close()

    # ---- 3. the sandbox: nothing sent; the payload printed only with debug ----
    for url, want in ((BASE + '/tests/?demo=1&via=kokua-kalihi&debug', True), (BASE + '/tests/?demo=1', False)):
        ctx = br.new_context(viewport={'width': 1440, 'height': 900}); sent = []; logs = []; rig(ctx, sent)
        page = ctx.new_page(); page.on('console', lambda m: logs.append((m.type, m.text))); page.goto(url, referer=BASE + '/track.html'); page.evaluate(IMPORT)
        r = page.evaluate("() => VL.logVisit('topics', 'view', { path: 'in' })"); page.wait_for_timeout(300)
        dbg = [l for l in logs if l[0] == 'debug' and 'first visit' in l[1]]
        ok(r is False and not [s for s in sent if s.get('p')] and (bool(dbg) == want), f'sandbox{" with debug" if want else ""}: nothing sent, payload printed: {bool(dbg)}')
        if want: ok(page.evaluate("() => VL.partnerWelcome('kokua-kalihi')") == 'Welcome, friends of Kokua Kalihi!' and not [s for s in sent if s.get('partners')], 'sandbox welcome line is a made-up sample, no request')
        if want: ok('ref_domain' not in (dbg[0][1] if dbg else ''), 'the page\'s own site is not a referrer')
        ctx.close()

    # ---- 4. no network at all: never throws, never retries ----
    ctx = br.new_context(viewport={'width': 800, 'height': 900}); sent = []
    ctx.route(re.compile(r'.*(supabase\.co|jsdelivr\.net/npm/@supabase).*'), lambda route: (sent.append(route.request.url), route.abort()))
    page = ctx.new_page(); errs = []; page.on('pageerror', lambda e: errs.append(str(e))); page.goto(BASE + '/tests/'); page.evaluate(IMPORT)
    r = page.evaluate("() => VL.logVisit('topics', 'view')"); page.wait_for_timeout(1500)
    w = page.evaluate("() => VL.partnerWelcome('kokua-kalihi')")
    dev = page.evaluate("() => JSON.parse(sessionStorage.getItem('hiphi_fv')).src.device")
    ok(r is False and w is None and not errs, f'offline: false and null, no error thrown ({len(sent)} attempts, no retry loop)')
    ok(len([u for u in sent if 'rpc/log_first_visit' in u]) <= 1, 'one attempt at most for the event')
    ok(dev == 'tablet', 'an 800px window is a tablet')
    ctx.close()
    br.close()
print(f'\n{passed} passed, {failed} failed')
sys.exit(1 if failed else 0)

File: pub/visitlog.js
// The first visit, counted privately (R-023 decision 8; backend migrations 067-068, docs/FIRST-VISIT-PLAN.md
// "Measurement"). One small row per screen event goes to log_first_visit, a database function anyone may call, which
// checks every value again and drops what does not fit. Staff read the weekly roll-up in Staff v2 (Outreach > Issues >
// First visit); nobody can read the rows themselves.
//
// What leaves the browser, and nothing else: a random id for this visit (sessionStorage, so it is gone when the tab
// closes; no cookie; never tied to an account), the screen and how it was left, seconds on it, the path, the lesson step
// and the "Try it" answer, counts (never an email, a name, an address or which stance), the ids of the issues followed,
// and where the person came from: ?via=, the utm_ words, and the referring site's name only (instagram.com, never the
// page). Phone, tablet or laptop comes from the window width, never the browser's identity string.
// Nothing at all is sent when the browser asks for Global Privacy Control, or from the sandbox (?demo=1), where the
// payload is printed to the console instead when the address also has `debug`.
//
//   logVisit(step, event, extra)  fire and forget: never throws, never waits, never retries; at most 60 a visit
//   visitVia()                    this visit's ?via= slug, or ''
//   partnerWelcome(slug)          the partner's welcome line (public_partners), or null; asked once per slug
import { DEMO, SUPABASE_URL, SUPABASE_KEY, supa } from './core.js';

const KEY = 'hiphi_fv', CAP = 60;
const STEPS = new Set(['topics', 'issues', 'stand', 'bill', 'session', 'hearing', 'you', 'soon', 'done', 'home', 'arrive', 'act', 'followask']);
const EVENTS = new Set(['view', 'next', 'skip', 'back', 'leave', 'done', 'answer']);
const SLUG = /^[a-z0-9-]{1,40}$/, UTM = /^[a-z0-9._-]{1,40}$/, SITE = /^[a-z0-9.-]{1,80}$/;
const DEBUG = /(^|[?&])debug(=|&|$)/.test(location.search);
const gpc = () => { try { return navigator.globalPrivacyControl === true; } catch { return false; } };

// ---- this visit: its id and where it came from, worked out once and kept for the life of the tab ----
let mem = null;   // when sessionStorage is not available (a private window that refuses it), this page load only
const read = () => { try { return JSON.parse(sessionStorage.getItem(KEY) || 'null') || mem; } catch { return mem; } };
const write = v => { mem = v; try { sessionStorage.setItem(KEY, JSON.stringify(v)); } catch { /* this page load only */ } };
function newId() {
  try { if (crypto.randomUUID) return crypto.randomUUID(); } catch { /* older browsers */ }
  const b = crypto.getRandomValues(new Uint8Array(16)); b[6] = (b[6] & 15) | 64; b[8] = (b[8] & 63) | 128;
  const h = [...b].map(x => x.toString(16).padStart(2, '0')).join('');
  return `${h.slice(0, 8)}-${h.slice(8, 12)}-${h.slice(12, 16)}-${h.slice(16, 20)}-${h.slice(20)}`;
}
// The words that say where someone came from: lower case, and anything that does not fit the pattern is dropped.
function fromUrl() {
  const q = new URLSearchParams(location.search), word = (k, re) => { const v = (q.get(k) || '').trim().toLowerCase(); return re.test(v) ? v : ''; };
  return { via: word('via', SLUG), utm_source: word('utm_source', UTM), utm_medium: word('utm_medium', UTM), utm_campaign: word('utm_campaign', UTM) };
}
// The referring site's name only, when it is another site: l.instagram.com and lm.facebook.com are the apps' link
// wrappers, m. and www. the same site again.
function refSite() {
  try {
    if (!document.referrer) return '';
    const u = new URL(document.referrer);
    if (!/^https?:$/.test(u.protocol) || u.hostname === location.hostname) return '';
    let h = u.hostname.toLowerCase();
    for (let k = 0; k < 3 && /^(www|m|l|lm)\./.test(h); k++) h = h.replace(/^(www|m|l|lm)\./, '');
    return SITE.test(h) ? h : '';
  } catch { return ''; }
}
const device = () => { const w = window.innerWidth || document.documentElement.clientWidth || 0; return w < 600 ? 'phone' : w < 1024 ? 'tablet' : 'laptop'; };
function visit() {
  let v = read();
  if (!v || !v.id) { v = { id: newId(), n: 0, src: { ...fromUrl(), ref_domain: refSite(), device: device() } }; write(v); }
  return v;
}
// Worked out as the page loads, before anything can change the address, so a ?via= is never lost. Nothing is sent.
if (!gpc()) { try { visit(); } catch { /* counting never gets in the way */ } }

// ---- what an event may carry, and only that ----
// A number out of its range is dropped, as the database drops it; only seconds are held to an hour, since a long stay
// on a screen is real.
const num = (x, lo, hi) => { if (typeof x !== 'number' || !Number.isFinite(x)) return undefined; const v = Math.round(x); return v >= lo && v <= hi ? v : undefined; };
function pick(step, event, x) {
  const out = {};
  if (['in', 'off', 'link'].includes(x.path)) out.path = x.path;
  const s = typeof x.seconds === 'number' && x.seconds > 3600 ? 3600 : num(x.seconds, 0, 3600); if (s !== undefined) out.seconds = s;
  const l = num(x.lesson_step, 0, 9); if (l !== undefined) out.lesson_step = l;
  if (['right', 'wrong', 'shown'].includes(x.quiz)) out.quiz = x.quiz;
  if (x.counts && typeof x.counts === 'object') {
    const c = {}, lim = { cats: 6, issues: 300, stances: 50 };
    for (const k of Object.keys(lim)) { const v = num(x.counts[k], 0, lim[k]); if (v !== undefined) c[k] = v; }
    for (const k of ['address', 'email']) if (typeof x.counts[k] === 'boolean') c[k] = x.counts[k];
    if (Object.keys(c).length) out.counts = c;
  }
  if (step === 'issues' && event === 'next' && Array.isArray(x.issue_ids)) out.issue_ids = [...new Set(x.issue_ids.filter(id => typeof id === 'string'))].slice(0, 300);
  return out;
}

// ---- sending: one request, no retry, no waiting ----
// "leave" is the tab closing (or the logo): the page may be gone before an ordinary request starts, so that one goes as a
// keepalive request, started at once, which the browser finishes after the page has closed. Only the public key goes
// with it: the function records no account either way.
function send(payload, leaving) {
  if (leaving) {
    return fetch(`${SUPABASE_URL}/rest/v1/rpc/log_first_visit`, { method: 'POST', keepalive: true, headers: { apikey: SUPABASE_KEY, 'Content-Type': 'application/json' }, body: JSON.stringify({ p: payload }) })
      .then(r => r.ok, () => false);
  }
  return supa().then(sb => sb.rpc('log_first_visit', { p: payload })).then(r => !r.error, () => false);
}

// Resolves to true when the row was sent, false when it was not (Global Privacy Control, the sandbox, the cap, a bad
// step or event, no network). Never rejects, so nobody has to catch it; nobody has to wait for it either.
export function logVisit(step, event, extra = {}) {
  try {
    if (gpc()) return Promise.resolve(false);
    if (!STEPS.has(step) || !EVENTS.has(event)) { console.warn('logVisit: not a first-visit step or event:', step, event); return Promise.resolve(false); }
    const v = visit();
    if (v.n >= CAP) return Promise.resolve(false);
    const payload = { visit: v.id, step, event };
    for (const [k, val] of Object.entries(v.src || {})) if (val) payload[k] = val;
    Object.assign(payload, pick(step, event, extra || {}));
    if (DEMO) { if (DEBUG) console.debug('first visit (sandbox, not sent):', payload); return Promise.resolve(false); }
    v.n += 1; write(v);
    return send(payload, event === 'leave').catch(() => false);
  } catch { return Promise.resolve(false); }
}

// This visit's partner slug (?via=), or '' when there is none. Read from the address itself under Global Privacy Control,
// where nothing is kept.
export function visitVia() {
  try { if (gpc()) return fromUrl().via; return visit().src?.via || ''; } catch { return ''; }
}

// The partner's welcome line ("Welcome, friends of ..."), or null: no such partner, no line, or no network. Asked once
// per slug per page load. The sandbox never asks the database: it shows a made-up line so the screen can be seen.
const welcomes = new Map();
export function partnerWelcome(slug) {
  const s = String(slug || '').toLowerCase();
  if (!/^[a-z0-9-]{2,40}$/.test(s)) return Promise.resolve(null);
  if (DEMO) return Promise.resolve(`Welcome, friends of ${s.split('-').map(w => w[0].toUpperCase() + w.slice(1)).join(' ')}!`);
  if (!welcomes.has(s)) welcomes.set(s, supa()
    .then(sb => sb.from('public_partners').select('welcome').eq('slug', s).maybeSingle())
    .then(r => (!r.error && r.data && r.data.welcome) || null, () => null));
  return welcomes.get(s);
}
